File: javascript/1160.Find-Words-That-Can-Be-Formed-by-Characters.js
/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
var countCharacters = function (words, chars) {
  let length = 0;

  const createMap = (word) => {
    const map = new Map();

    for (const c of word) {
      map.set(c, (map.get(c) || 0) + 1);
    }

    return map;
  };

  const hasAllLetters = (wordMap) => {
    for (const [key, val] of wordMap) {
      if (!charMap.has(key) || charMap.get(key) < val) {
        return false;
      }
    }

    return true;
  };

  const charMap = createMap(chars);

  for (const word of words) {
    if (hasAllLetters(createMap(word))) {
      length += word.length;
    }
  }

  return length;
};

File: python/1160.Find-Words-That-Can-Be-Formed-by-Characters.py
class Solution:
    def hasAllChars(self, charsMap, wordMap) -> bool:
        for key, val in wordMap.items():
            if key not in charsMap or charsMap.get(key) < val:
                return False
        return True

    def createWordMap(self, word) -> Dict[str, int]:
        m = {}

        for w in word:
            m[w] = m.get(w, 0) + 1

        return m

    def countCharacters(self, words: List[str], chars: str) -> int:
        charsMap = self.createWordMap(chars)
        length = 0
        
        for word in words:
            if (self.hasAllChars(charsMap, self.createWordMap(word))):
                length += len(word)

        return length